receiver/sapnetweaverreceiver/scraper.go

```diff
@@ -0,0 +1,148 @@
+	soapClient, err := newSoapClient(s.cfg, host, s.settings)
+	if err != nil {
+		return fmt.Errorf("failed to create client: %w", err)
+	}
+	s.client = soapClient
+	s.service = newWebService(s.client)
+
+	return nil
+}
+
+func (s *sapNetweaverScraper) GetCurrentInstance() error {
+	var response *models.GetInstancePropertiesResponse
+	response, err := s.service.GetInstanceProperties()
+	if err != nil {
+		return err
+	}
+
+	for _, prop := range response.Properties {
+		switch prop.Property {
+		case "INSTANCE_NAME":
+			s.instance = prop.Value
+		case "SAPLOCALHOST":
+			s.hostname = prop.Value
+		}
+	}
+
+	return nil
+}
+
+func (s *sapNetweaverScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	if s.client == nil || s.service == nil {
+		return pmetric.Metrics{}, errors.New("failed to create client")
+	}
+
+	errs := &scrapererror.ScrapeErrors{}
+	err := s.GetCurrentInstance()
+	if err != nil {
+		errs.AddPartial(1, fmt.Errorf("failed to get current instance details: %w", err))
+	}
+
+	s.collectMetrics(ctx, errs)
+	return s.mb.Emit(), errs.Combine()
+}
+
+func (s *sapNetweaverScraper) collectMetrics(ctx context.Context, errs *scrapererror.ScrapeErrors) {
+	now := pcommon.NewTimestampFromTime(time.Now())
+	s.collectAlertTree(ctx, now, errs)
+	s.collectEnqGetLockTable(ctx, now, errs)
+}
+
+func (s *sapNetweaverScraper) collectAlertTree(_ context.Context, now pcommon.Timestamp, errs *scrapererror.ScrapeErrors) {
+	alertTreeResponse := map[string]string{}
+	alertTree, err := s.service.GetAlertTree()
+	if err != nil {
+		errs.AddPartial(1, fmt.Errorf("failed to collect Alert Tree metrics: %w", err))
+		return
+	}
+
+	for _, node := range alertTree.AlertNode {
+		value := strings.Split(node.Description, " ")
+		alertTreeResponse[node.Name] = value[0]
+		if node.Name == "ICM" {
+			alertTreeResponse[node.Name] = string(node.ActualValue)
+		}
+	}
+
+	s.recordSapnetweaverHostCPUUtilizationDataPoint(now, alertTreeResponse, errs)
+	s.recordSapnetweaverHostMemoryVirtualOverheadDataPoint(now, alertTreeResponse, errs)
+	s.recordSapnetweaverHostMemoryVirtualSwapDataPoint(now, alertTreeResponse, errs)
+	s.recordSapnetweaverSessionsHTTPCountDataPoint(now, alertTreeResponse, errs)
+	s.recordCurrentSecuritySessions(now, alertTreeResponse, errs)
+	s.recordSapnetweaverWorkProcessesActiveCount(now, alertTreeResponse, errs)
+	s.recordSapnetweaverSessionsWebCountDataPoint(now, alertTreeResponse, errs)
+	s.recordSapnetweaverSessionsBrowserCountDataPoint(now, alertTreeResponse, errs)
+	s.recordSapnetweaverSessionsEjbCountDataPoint(now, alertTreeResponse, errs)
+	s.recordSapnetweaverIcmAvailabilityDataPoint(now, alertTreeResponse, errs)
+	s.recordSapnetweaverHostSpoolListUsedDataPoint(now, alertTreeResponse, errs)
+	s.recordSapnetweaverShortDumpsCountDataPoint(now, alertTreeResponse, errs)
+	s.mb.EmitForResource(metadata.WithSapnetweaverInstance(s.instance), metadata.WithSapnetweaverNode(s.hostname))
```

Comment:
I notice we emit for the same resource in both `collectEnqGetLockTable` and `collectAlertTree` - can we hoist this up into the calling function, so it's only done once? A receiver should only emit once for any particular resource, generally.

Reply:
Put a single emit in the collect `collectMetrics` method